# 프로미스
ES6에서 비동기 처리를 위한 패턴으로 프로미스를 도입했다. 전통적인 콜백 패턴이 가진 단점을 보완하며 비동기 처리 시점을 명확하게 표현할 수 있다는 장점이 있다.
## 1. 비동기 처리를 위한 콜백 패턴의 단점
1) 콜백 헬
___

    const get = url => {
        const xhr = new XMLHttpRequest();
        xhr.open('GET', url);  
        xhr.send();
        xhr.onload = () => {
            if(xhr.status === 200){
                console.log(JSON.parse(xhr.response));
            }else{
                console.error(`${xhr.status} ${xhr.statusText}`);
            }
        };
    };

    get('https://jsonplaceholder.typicode.com/posts/1');

위의 get 함수는 비동기 함수로 함수 내부의 비동기로 동작하는 코드가 완료되지 않아도 기다리지 않고 즉시 종료 된다. 즉 함수 내부의 비동기로 동작하는 코드는 비동기 함수가 종료된 이후에 완료된다.

    let g = 0;

    //setTimeout()은 비동기 함수로 콜백함수의 결과를 외부로 반환하거나 상위 스코프 변수에 할당X
    setTimeout(() => { g=100;}, 0);
    console.log(g); //0

위 예제에서 get 함수가 비동기 함수인 이유는 onload 이벤트 핸들러가 비동기로 동작하기 때문이다.

    const todus;
    const get = url => {
        const xhr = new XMLHttpRequest();
        xhr.open('GET', url);  
        xhr.send();
        xhr.onload = () => {
            if(xhr.status === 200){
                todus = JSON.parse(xhr.response);
            }else{
                console.error(`${xhr.status} ${xhr.statusText}`);
            }
        };
    };

    const resp = get('https://jsonplaceholder.typicode.com/posts/1');
    console.log(resp);  // undefined
    //onload 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러는 언제나 console.log가 종료한 이후에 호출되기 때문에 결과를 읽어올 수 없다.

서버로부터 응답이 도착하면 xhr 객체에서 load 이벤트가 발생한다. onload 이벤트 핸들러는 load 이벤트가 발생하면 태스크 큐에 저장되어 대기하다가 콜 스택이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다. 따라서 get 함수 이후에 consoloe.log가 100번 호출된다 하더라도 onload 이벤트 핸들러는 모든 console.log가 종료한 이후에 실행된다.

이처럼 비동기 함수는 비동기 처리 결과를 외부에 반환할 수 없고, 상위 스코프의 변수에 할당할 수도 없다. 그래서 서버의 응답에 대한 후속 처리는 비동기 함수 내부에서 수행해야 한다.

비동기 처리 결과에 대한 후속 처리를 수행하는 비동기 함수가 처리 결과를 가지고 또 비동기 함수를 호출하는 콜백함수의 중첩이 복잡도를 증가 시키는데 이를 '콜백 헬' 이라고 한다.

콜백 헬 사례

    get(`${url}/posts/1`, ({userId}) => {
        console.log(userId);  //1

        get(`${url}/users/${userId}`, userInfo => {
            console.log(userInfo);
        });
    });

    get('/step1', a => {
        get(`/step2/${a}`, b => {
            get(`/step3/${b}`, c => {
                get(`/step4/${c}`, d => {
                    console.log(d);
                });
            });
        });
    });

2) 에러처리의 한계
___

    try {
        setTimeout(( => { throw new Error('Error!');}, 1000));
    } catch(e) {
        console.error('캐치한 에러', e);
    }

에러는 호출자 방향으로 전파되는데 setTimeout 함수의 콜백 함수를 호출한 것이 setTimeout 함수가 아니기 때문에 try 절에서 에러 발생을 캐치하지 못한다.

## 2. 프로미스 생성
Promise 생성자 함수를 사용하여 프로미스 객체를 생성한다. 이는 호스트 객체가 아니라 표준 빌트인 객체다. Promise 생성자 함수는 비동기 처리를 수행할 콜백함수를 인수로 전달 받는데 이 콜백 함수는 resolve, reject 함수이다.

    const promise = new Promise((resolve, reject) => {
        if(/* 비동기 처리 성공 */)
            resolve('result');
        else /* 비동기 처리 실패 */
            reject('failure reason')'
    });

    const promiseGet = url => {
        return new Promise((resolve, reject) => {
            const xhr = new XMLHttpRequest();
            xhr.open('GET', url);  
            xhr.send();
            xhr.onload = () => {
                if(xhr.status === 200){
                    resolve(JSON.parse(xhr.response));
                }else{
                    reject(new Error(xhr.status));
                }
            };
        });
    };

    promiseGet('https://jsonplaceholder.typicode.com/posts/1');

프로미스는 다음과 같이 상태 정보를 갖는다.
* pending : 프로미스가 생성된 직후, 비동기 처리가 아직 수행되지 않은 상태
* fulfilled : resolve 함수 호출, 비동기 처리가 수행되어 성공한 상태
* rejected : reject 함수 호출, 비동기 처리가 수행되어 실패한 상태

pending이 아닌 상태로 비동기 처리가 수행된 상태를 'settled' 상태라고 한다. settled 상태가 되면 다른 상태로 변화할 수 없다.

## 3. 프로미스의 후속 처리 메서드
비동기 처리 상태가 변화하면 이에 따른 후속 처리를 해야하는데 이를 위해 프로미스는 후속 메서드 then, catch, finally를 제공한다. 모든 후속처리 메서드는 프로미스를 반환하며 비동기로 동작한다.

__then__ : 두 개의 콜백 함수를 인수로 전달 받는다. 첫 번째는 resolve 함수가 호출된 상태에 비동기 처리 결과를 인수로 전달 받고, 두 번째는 reject 함수가 호출된 상태에 에러를 인수로 전달받는다.

    new Promise(resolve => reoleve('fulfilled'))
        .then(v => console.log(v), e => console.error(e));  // fulfilled

    new Promise((_, reject) => reject(new Error('rejected')))
        .then(v => console.log(v), e => console.error(e));  //Error : rejected

__catch__ : 한 개의 콜백 함수를 인수로 전달받음, 프로미스가 rejected 상태인 경우만 호출된다. then과 동일하게 동작

__finally__ : 프로미스의 성공, 실패여부와 상관없이 무조건 한 번 호출된다. 따라서 프로미스의 상태와 상관없이 공통적으로 수행해야할 처리 내용이 있을 때 유용하다.

    const promiseGet = url => {
        return new Promise((resolve, reject) => {
            const xhr = new XMLHttpRequest();
            xhr.open('GET', url);  
            xhr.send();
            xhr.onload = () => {
                if(xhr.status === 200){
                    resolve(JSON.parse(xhr.response));
                }else{
                    reject(new Error(xhr.status));
                }
            };
        });
    };

    promiseGet('https://jsonplaceholder.typicode.com/posts/1')
        .then(res => console.log(res))
        .catch(err => console.error(err))
        .finally(() => console.log('Bye!'));


